### PR TITLE
Add support for interconversion from data to gsd

### PR DESCRIPTION
**Pull request by clpetix**
Addresses Issue #36

### CODE REVIEW

File: src/lammpsio/snapshot.py [hunk 1]
```diff
@@ -176,19 +183,90 @@ def to_hoomd_gsd(self, type_map=None):
...
+                    for typeidx, typeid in enumerate(sorted_typeids):
+                        frame.bonds.typeid[self.bonds.typeid == typeid] = typeidx
+                else:
+                    frame.bonds.types = list(bond_type_map.values())
+                    reverse_bond_type_map = {
+                        typeid: typeidx
+                        for typeidx, typeid in enumerate(bond_type_map.keys())
+                    }
+                    for i, typeid in enumerate(self.bonds.typeid):
+                        frame.bonds.typeid[i] = reverse_bond_type_map[typeid]
```
Comment (clpetix) [inline src/lammpsio/snapshot.py:204]: I understand why this is necessary, but should we make a method or function to do this rather than duplicating the code over and over?

File: src/lammpsio/snapshot.py [hunk 2]
```diff
@@ -176,19 +183,90 @@ def to_hoomd_gsd(self, type_map=None):
             frame.particles.mass = self.mass.copy()
         if self.has_molecule():
             frame.particles.body = self.molecule - 1
-
+        if self.has_bonds():
+            frame.bonds.N = self.bonds.N
+            frame.bonds.types = self.bonds.id
+            frame.bonds.group = self.bonds.members - 1
```
Comment (clpetix) [inline src/lammpsio/snapshot.py:189]: Is this acceptable to just subtract one to move from LAMMPS one-index and HOOMD's zero-index?  I couldn't think of any scenarios where this wouldn't work, but wanted to check.

File: src/lammpsio/snapshot.py [hunk 3]
```diff
@@ -106,7 +106,14 @@ def from_hoomd_gsd(cls, frame):
 
         return snap, type_map
 
-    def to_hoomd_gsd(self, type_map=None):
+    def to_hoomd_gsd(
+        self,
+        type_map=None,
```
Comment (clpetix) [inline src/lammpsio/snapshot.py:111]: If we stay with this, we should note to consider renaming`type_map` to `particle_type_map` just to be clear.

Comment (mphoward) [reply src/lammpsio/snapshot.py:111]: This will be an issue for `from_hoomd_gsd` as well since that returns a `type_map`. We could do like you've done here. We could also accept a dict of dicts (keyed on "particles", "bonds", etc.), and for backwards compatibility, convert a regular dict to the particles key. Or, we could create a typemap object that stores all of these, and require this be passed when mapping topology data (and internally convert the dict of particle types to this object).

File: src/lammpsio/snapshot.py (hunk 2 above)
Comment (mphoward) [reply src/lammpsio/snapshot.py:189]: It should be, this would be how we implicitly shift the `id`.

File: src/lammpsio/snapshot.py (hunk 1 above)
Comment (mphoward) [reply src/lammpsio/snapshot.py:204]: Yes, I agree it would be nice to have a helper function that does this.

File: src/lammpsio/snapshot.py [hunk 4]
```diff
@@ -612,3 +680,29 @@ def reorder(self, order, check_order=True):
             self._charge = self._charge[order]
         if self.has_mass():
             self._mass = self._mass[order]
+
+    def _set_type_id(gsd_typeid, lammps_typeid, type_map):
```
Comment (clpetix) [inline src/lammpsio/snapshot.py:684]: add self

**Comment by clpetix** [comment]
@mphoward is this what you had in mind for the TypeMap object? I think this is a working implementation of the conversion to GSD. I should be able to add the conversion from gsd and some unit tests pretty easily from here!

**Comment by clpetix** [comment]
@mphoward could you give this a quick look and check that this is how you wanted LabelMap implemented?

If so, I will work on tests and getting everything polished to go!

File: src/lammpsio/snapshot.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def __init__(self, N, box, step=None, num_types=None):
         self._impropers = None
 
     @classmethod
-    def from_hoomd_gsd(cls, frame):
+    def from_hoomd_gsd(cls, frame, type_map=None):
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:50]: I think this is a stray argument, right?

File: src/lammpsio/snapshot.py [hunk 6]
```diff
@@ -91,22 +91,77 @@ def from_hoomd_gsd(cls, frame):
         snap.charge = frame.particles.charge
         snap.mass = frame.particles.mass
 
-        snap.molecule = frame.particles.body + 1
+        if frame.particles.body is not None:
+            snap.molecule = frame.particles.body + 1
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:95]: Was this causing an issue? The default value of `body` should be -1 in HOOMD for something that doesn't have such a label, so it would become 0 meaning "not a molecule" to LAMMPS. I would have assumed that this if statement always evaluates True.

If the goal is not to write molecule info when it isn't present in the GSD file, we could be more sophisticated and check if `np.any(frame.particles.body != -1)`, then only do this molecule assigning and validation of the values, possibly also auto-remapping negative values of the body if they are all negative.

File: src/lammpsio/snapshot.py [hunk 7]
```diff
@@ -91,22 +91,77 @@ def from_hoomd_gsd(cls, frame):
...
-            or frame.angles.N > 0
-            or frame.dihedrals.N > 0
-            or frame.impropers.N > 0
-        ):
-            warnings.warn("Conversion of topology from gsd is not supported")
-        type_map = {typeid + 1: i for typeid, i in enumerate(frame.particles.types)}
+        if frame.bonds.N != 0:
+            snap.bonds = Bonds(N=frame.bonds.N)
+            snap.bonds.id = frame.bonds.typeid
+            snap.bonds.members = frame.bonds.group + 1
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:102]: I would move the labelmap logic for each of these bonding types up here, and only assign the labels if they exist. If there are bonds in the GSD file, there has to be at least one type stored.

File: src/lammpsio/snapshot.py [hunk 8]
```diff
@@ -91,22 +91,77 @@ def from_hoomd_gsd(cls, frame):
...
+            snap.impropers.id = frame.impropers.typeid
+            snap.impropers.members = frame.impropers.group + 1
+
+        # set particle label
+        label_map_particle = None
+        if frame.particles.types is not None:
+            label_map_particle = {
+                typeid + 1: i for typeid, i in enumerate(frame.particles.types)
+            }
+        snap.label = LabelMap(map=label_map_particle)
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:125]: This would go best above (near where the old `type_map` was deleted), before dealing with the bonded interactions.

File: src/lammpsio/snapshot.py [hunk 9]
```diff
@@ -91,22 +91,77 @@ def from_hoomd_gsd(cls, frame):
...
+
+        # set improper label
+        label_map_improper = None
+        if frame.impropers.types is not None:
+            label_map_improper = {
+                typeid + 1: i for typeid, i in enumerate(frame.impropers.types)
+            }
+        frame.impropers.label = LabelMap(map=label_map_improper)
 
         return snap, type_map
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:159]: Make sure that a `type_map` dict is still generated from the particle label map, for backward compatibility.

File: src/lammpsio/snapshot.py [hunk 10]
```diff
@@ -164,31 +219,114 @@ def to_hoomd_gsd(self, type_map=None):
...
                     frame.particles.typeid[self.typeid == typeid] = typeidx
             else:
-                frame.particles.types = list(type_map.values())
-                reverse_type_map = {
-                    typeid: typeidx for typeidx, typeid in enumerate(type_map.keys())
-                }
-                for i, typeid in enumerate(self.typeid):
-                    frame.particles.typeid[i] = reverse_type_map[typeid]
+                if isinstance(type_map, dict):
+                    self.label = LabelMap(map=type_map)
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:223]: Here we don't want to actually override the value of `label`. In general, we also probably want to deprecate setting the `type_map` value entirely, as this should all be handled through the snapshot. Something like:
```py
if type_map is not None:
    warnings.warn(...)
    type_label_map = LabelMap(type_map)
else:
    type_label_map = self.label
```
then use this in your `_set_type_id` method

File: src/lammpsio/snapshot.py [hunk 11]
```diff
@@ -164,31 +219,114 @@ def to_hoomd_gsd(self, type_map=None):
...
+                    frame.bonds.types = self.bonds.label.types
+                    Snapshot._set_type_id(
+                        self,
+                        self.bonds.typeid,
+                        frame.bonds.typeid,
+                        self.bonds.label,
+                    )
+                else:
+                    sorted_typeids = numpy.sort(numpy.unique(self.bonds.typeid))
+                    frame.bonds.types = [str(typeid) for typeid in sorted_typeids]
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:258]: I think it would be clearer & reduce duplication to generate a `LabelMap` from these lines, then just use `_set_type_id`.

**Review by mphoward — COMMENTED** [review]
Overall this looks great! Here are some comments, but then please move forward with testing.

File: src/lammpsio/snapshot.py (hunk 6 above)
Comment (clpetix) [reply src/lammpsio/snapshot.py:95]: Not sure what I was doing here. I will revert this change and we can address this later, if needed.

Comment (clpetix) [reply src/lammpsio/snapshot.py:95]: I figured out why I was doing this. I get this error when using `to_hoomd_gsd` then going back with `from_hoomd_gsd`: 
``` 
snap.molecule = frame.particles.body + 1
TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'
```

Comment (mphoward) [reply src/lammpsio/snapshot.py:95]: Hmm, so which direction is causing the error, or what is the test case? I do see that it seems like the default value of any attribute in a `gsd.hoomd.Frame` is `None` if it has not been assigned. I'm not sure what happens when read from file though:
```py
>>> import gsd.hoomd
>>> frame = gsd.hoomd.Frame()
>>> print(frame.particles.body)
None
>>> print(frame.particles.position)
None
```

Comment (clpetix) [reply src/lammpsio/snapshot.py:95]: The error pops up when using `from_hoomd_gsd` to read a GSD where body was never set. In lammpsio, it only gets set in `to_hoomd_gsd` here:
```
if self.has_molecule():
            frame.particles.body = self.molecule - 1
```

So when I was using lammpsio to go to GSD then back to LAMMPS data file, it causes as error.  I think we just need to check if it is not none before trying to set `snap.molecule`.

Comment (mphoward) [reply src/lammpsio/snapshot.py:95]: Is this done by passing through a GSD file, though, or on the frame object directly?

I think checking for `None` is probably the right behavior, but I wanted to make sure I understood what was triggering it.

Comment (clpetix) [reply src/lammpsio/snapshot.py:95]: I was passing the frame directly and it caused the error. I just checked writing to file and then passing a GSD file and that does not trigger the error. The body values default to -1.

Comment (mphoward) [reply src/lammpsio/snapshot.py:95]: Got it, thanks!

I think that `from_hoomd_gsd` should work regardless of whether the `Frame` was created directly in Python or read from a GSD file, so I think this means that we actually need to do the `is not None` check on _every_ property before we assign it to the lammpsio `Snapshot`.

Similarly, on the `to_hoomd_gsd` side, we should only set the attributes of the GSD `Frame` if `has_X() is True` in the `Snapshot`. lammpsio does not have a default for attributes that are not set (just a default for how they are initialized), so I think we should just let GSD do its thing with those.

**Comment by clpetix** [comment]
@mphoward, this PR should also be ready for a review!

File: src/lammpsio/snapshot.py [hunk 12]
```diff
@@ -107,18 +107,64 @@ def from_hoomd_gsd(cls, frame):
...
-        return snap, type_map
-
-    def to_hoomd_gsd(self, type_map=None):
+        # set particle label
+        label_map_particle = None
+        if frame.particles.types is not None:
+            label_map_particle = {
+                typeid + 1: i for typeid, i in enumerate(frame.particles.types)
+            }
+        snap.label = LabelMap(map=label_map_particle)
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:116]: I think this should only be set if the types are present (which they are basically all the time)
```suggestion
            snap.label = LabelMap(map=label_map_particle)
```

File: src/lammpsio/snapshot.py [hunk 13]
```diff
@@ -107,18 +107,64 @@ def from_hoomd_gsd(cls, frame):
...
+            snap.impropers.typeid = frame.impropers.typeid + 1
+            snap.impropers.members = frame.impropers.group + 1
+
+            # set improper label
+            label_map_improper = {
+                typeid + 1: i for typeid, i in enumerate(frame.impropers.types)
+            }
+            snap.impropers.label = LabelMap(map=label_map_improper)
+
+        return snap, snap.label
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:162]: This should still be a dictionary of particle types for backwards compatibility
```suggestion
        return snap, label_map_particle
```

File: src/lammpsio/snapshot.py [hunk 14]
```diff
@@ -176,31 +222,122 @@ def to_hoomd_gsd(self, type_map=None):
...
             else:
-                frame.particles.types = list(type_map.values())
-                reverse_type_map = {
-                    typeid: typeidx for typeidx, typeid in enumerate(type_map.keys())
-                }
-                for i, typeid in enumerate(self.typeid):
-                    frame.particles.typeid[i] = reverse_type_map[typeid]
+                if type_map is not None:
+                    warnings.warn(
+                        "type_map will be deprecated, use label instead.",
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:227]:
```suggestion
                        "type_map is deprecated, use label instead.",
```

File: src/lammpsio/snapshot.py [hunk 15]
```diff
@@ -176,31 +222,122 @@ def to_hoomd_gsd(self, type_map=None):
...
+                        DeprecationWarning,
+                        stacklevel=2,
+                    )
+                    type_label_map = LabelMap(map=type_map)
+                else:
+                    type_label_map = self.label
+                frame.particles.types = type_label_map.types
+                Snapshot._set_type_id(
+                    self,
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:236]: Is there a reason this doesn't work?
```suggestion
                self._set_type_id(
```

If you aren't really using any class variables inside this method, you could also define it outside the class as a private helper function.

File: src/lammpsio/snapshot.py [hunk 16]
```diff
@@ -176,31 +222,122 @@ def to_hoomd_gsd(self, type_map=None):
...
+                    self.typeid,
+                    frame.particles.typeid,
+                    self.label,
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:239]: Does this need to be this?
```suggestion
                    type_label_map,
```

Make sure there is a unit test that covers this case.

File: src/lammpsio/snapshot.py [hunk 17]
```diff
@@ -176,31 +222,122 @@ def to_hoomd_gsd(self, type_map=None):
...
+                )
         if self.has_charge():
             frame.particles.charge = self.charge.copy()
         if self.has_mass():
             frame.particles.mass = self.mass.copy()
         if self.has_molecule():
             frame.particles.body = self.molecule - 1
+        if self.has_bonds():
+            frame.bonds.N = self.bonds.N
+            frame.bonds.types = self.bonds.id
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:249]: I'm confused by this. It doesn't seem like we should assign `bonds.N` types as the default behavior.

If bonds has typeids, then you have two cases:
* There is a label map, so the types come from the label map.
* There is no label map, so the types are the unique typeids.

If bonds does not have typeids, you should not write them. It will be an ill-formed GSD file, but that is not our problem.

File: src/lammpsio/topology.py [hunk 18]
```diff
@@ -150,6 +153,21 @@ def num_types(self, value):
         else:
             self._num_types = None
 
+    @property
+    def label(self):
```
Comment (mphoward) [inline src/lammpsio/topology.py:157]: I would prefer to call these properties `type_label` or `typeid_label` to be consistent with the LAMMPS nomenclature. LAMMPS calls its numerical values `type`, but we use `typeid` to make clear that it is indeed just a number.

Also, is the snapshot missing a similar property?

File: src/lammpsio/topology.py [hunk 19]
```diff
@@ -241,3 +259,49 @@ class Impropers(Topology):
 
     def __init__(self, N, num_types=None):
         super().__init__(N=N, num_members=4, num_types=num_types)
+
+
+class LabelMap(MutableMapping):
+    def __init__(self, map=None):
+        self._map = {}
+        if map is not None:
+            self._map.update(map)
```
Comment (mphoward) [inline src/lammpsio/topology.py:268]: I think you can just invoke `self.update` directly. That way you rely on the setters to do everything.
```suggestion
        if map is not None:
            self.update(map)
```

File: src/lammpsio/topology.py [hunk 20]
```diff
@@ -1,3 +1,5 @@
+from collections.abc import MutableMapping
```
Comment (mphoward) [inline src/lammpsio/topology.py:1]: I prefer not to import external classes into our namespace.

```suggestion
import collections.abc
```

File: src/lammpsio/topology.py [hunk 21]
```diff
@@ -241,3 +259,49 @@ class Impropers(Topology):
 
     def __init__(self, N, num_types=None):
         super().__init__(N=N, num_members=4, num_types=num_types)
+
+
+class LabelMap(MutableMapping):
```
Comment (mphoward) [inline src/lammpsio/topology.py:264]:
```suggestion
class LabelMap(collections.abc.MutableMapping):
```

File: src/lammpsio/snapshot.py [hunk 22]
```diff
@@ -107,18 +107,64 @@ def from_hoomd_gsd(cls, frame):
...
-    def to_hoomd_gsd(self, type_map=None):
+        # set particle label
+        label_map_particle = None
+        if frame.particles.types is not None:
+            label_map_particle = {
+                typeid + 1: i for typeid, i in enumerate(frame.particles.types)
+            }
+        snap.label = LabelMap(map=label_map_particle)
+
+        if frame.bonds.N != 0:
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:118]: What is going to happen in these cases if one or more of these attributes of `frame.bonds` is None? Is that possible?

**Review by mphoward — CHANGES_REQUESTED** [review]
This is a good draft! Here are some comments that need investigating.

File: src/lammpsio/snapshot.py (hunk 15 above)
Comment (clpetix) [reply src/lammpsio/snapshot.py:236]: I think you're write that a private helper function would be best here to not clutter the class.

File: src/lammpsio/snapshot.py (hunk 17 above)
Comment (clpetix) [reply src/lammpsio/snapshot.py:249]: That makes sense. I think this should raise an error letting users know that they must specify typids to convert to GSD.

File: src/lammpsio/snapshot.py (hunk 22 above)
Comment (clpetix) [reply src/lammpsio/snapshot.py:118]: This is happy to just have it be none but it would obviously cause issues. Should we add like a private `_interoperable()` method or something to the Topology class that checks that it has members, typeid, and id and raise an error that it isn't convertable without all of those properties being defined?

File: src/lammpsio/snapshot.py (hunk 17 above)
Comment (mphoward) [reply src/lammpsio/snapshot.py:249]: I don’t think it has to be an error since it returns a Frame that the caller could decide to modify. GSD should error out itself if a bad Frame gets written to file

Comment (clpetix) [reply src/lammpsio/snapshot.py:249]: Should it be a warning? I feel like it could be pretty annoying to try to debug why your topology isn't converting while everything seemingly works fine.

Comment (mphoward) [reply src/lammpsio/snapshot.py]: What about forcing a call to `frame.validate()`? That way GSD will internally self-validate:

https://gsd.readthedocs.io/en/v3.3.2/python-module-gsd.hoomd.html#gsd.hoomd.Frame.validate

I think this method ignores things that are `None`, so that seems to be allowed behavior and I think we should respect that.

**Comment by clpetix** [comment]
@mphoward, thanks for the review! I think that I have addressed all of your comments whenever you get a chance to look at this again!

File: src/lammpsio/snapshot.py [hunk 23]
```diff
@@ -171,36 +230,128 @@ def to_hoomd_gsd(self, type_map=None):
...
 
-        if (
-            self.has_bonds()
-            or self.has_angles()
-            or self.has_dihedrals()
-            or self.has_impropers()
-        ):
-            warnings.warn("Conversion of topology to gsd is not supported")
-
+        frame.validate()
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:354]: I don't think a validation is necessary when going _to_ GSD. Maybe in the _from_ direction to sanitize our inputs.

**Review by mphoward — CHANGES_REQUESTED** [review]
Thanks! I think some of the comments may have gotten a little confused in the discussion. I tried to clarify here, but let me know if you are still unsure about anything!

File: src/lammpsio/snapshot.py [hunk 24]
```diff
@@ -171,36 +230,128 @@ def to_hoomd_gsd(self, type_map=None):
...
+                    for typeid in sorted_typeids:
+                        bond_label_map[typeid] = str(typeid)
+                    bond_label_map = LabelMap(map=bond_label_map)
+                # set bond types & typeid
+                frame.bonds.types = bond_label_map.types
+                _set_type_id(
+                    self.bonds.typeid,
+                    frame.bonds.typeid,
+                    bond_label_map,
+                )
```
[the remark this answers is not in the document]
Comment (clpetix) [reply src/lammpsio/snapshot.py]: I am a bit confused on this. 

Why are you setting the `frame.bonds.typeid` to zeros? Also this is setting `bond_label_map = self.bonds.type_label` and then overwriting it? 

And `_set_type_id` is a helper function that maps from LAMMPS type IDs to HOOMD type IDs but doesn't return anything. If it requires a bond label map as an input and doesn't modify it, I'm confused why it would also set it.

[the remark this answers is not in the document]
Comment (mphoward) [reply src/lammpsio/snapshot.py:317]: > Why are you setting the `frame.bonds.typeid` to zeros?

I think your code requires that `frame.bonds.typeid` have something in it before you can call `_set_type_id` so this a temporary placeholder. It could also use `numpy.ones` or `numpy.empty`. You were filling `frame.bonds.typeid` with `self.bonds.id` above (i.e., 1, 2, ... N), which didn't make sense because that would imply each bond is its own type.
 
> Also this is setting `bond_label_map = self.bonds.type_label` and then overwriting it?
>
> And `_set_type_id` is a helper function that maps from LAMMPS type IDs to HOOMD type IDs but doesn't return anything. If it requires a bond label map as an input and doesn't modify it, I'm confused why it would also set it.

If `self.bonds.type_label is not None`, `bond_label_map` will be the label map, otherwise it will default to `None`. If `self.bonds.has_typeid() is True`, then `_set_type_id` will be called with `bond_label_map`. My comment above said that you should modify `_set_type_id` so that it creates the 1, 2, 3, ... default label map internally when it is given a label map that is None (to reduce code duplication) and returns the label map it actually used. That would mean that if `bond_label_map is not None`, the assignment does nothing; if it is None, it will return the auto-generated map. It is still possible for `bond_label_map` to be None if it was not set and there were no typeids.

[the remark this answers is not in the document]
Comment (clpetix) [reply src/lammpsio/snapshot.py:317]: Ah I see, I had missed the bit about modifying `_set_type_id` to create `bond_label_map` if it is given `None`. That all makes sense! Thanks!

**Comment by clpetix** [comment]
@mphoward I implemented the suggested changes!

File: src/lammpsio/snapshot.py [hunk 25]
```diff
@@ -170,37 +258,84 @@ def to_hoomd_gsd(self, type_map=None):
...
         if self.has_typeid():
             frame.particles.typeid = numpy.zeros(self.N, dtype=int)
-            if type_map is None:
-                sorted_typeids = numpy.sort(numpy.unique(self.typeid))
-                frame.particles.types = [str(typeid) for typeid in sorted_typeids]
-                for typeidx, typeid in enumerate(sorted_typeids):
-                    frame.particles.typeid[self.typeid == typeid] = typeidx
+            if type_map is not None:
+                warnings.warn(
+                    "type_map is deprecated, use label instead.",
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:263]:
```suggestion
                    "type_map is deprecated, use Snapshot.type_label instead.",
```

File: src/lammpsio/snapshot.py [hunk 26]
```diff
@@ -170,37 +258,84 @@ def to_hoomd_gsd(self, type_map=None):
...
+                type_label_map = LabelMap(map=type_map)
             else:
-                frame.particles.types = list(type_map.values())
-                reverse_type_map = {
-                    typeid: typeidx for typeidx, typeid in enumerate(type_map.keys())
-                }
-                for i, typeid in enumerate(self.typeid):
-                    frame.particles.typeid[i] = reverse_type_map[typeid]
+                type_label_map = self.type_label
+            if self.has_typeid():
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:270]: These `if`s are a little off here (you have `self.has_typeid()` nested inside itself). I think you need to rejigger this a little to get the type label map before checking the typeid, more like how it's done below for the topology.

File: src/lammpsio/snapshot.py [hunk 27]
```diff
@@ -170,37 +258,84 @@ def to_hoomd_gsd(self, type_map=None):
...
+                    self.typeid, frame.particles.typeid, type_label_map
+                )
+            frame.particles.types = type_label_map.types
         if self.has_charge():
             frame.particles.charge = self.charge.copy()
         if self.has_mass():
             frame.particles.mass = self.mass.copy()
         if self.has_molecule():
             frame.particles.body = self.molecule - 1
+        if self.has_bonds():
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:282]: I forget the details of this method, but iirc, I think this should be:
```suggestion
        if self.bonds is not None:
```
because `has_bonds()` also checks for N > 0, but it is allowed to be 0 now.

File: src/lammpsio/snapshot.py [hunk 28]
```diff
@@ -170,37 +258,84 @@ def to_hoomd_gsd(self, type_map=None):
...
+            bond_label_map = self.bonds.type_label
+            if self.bonds.has_typeid():
+                frame.bonds.typeid = numpy.zeros(self.bonds.N, dtype=int)
+                bond_label_map = _set_type_id(
+                    self.bonds.typeid,
+                    frame.bonds.typeid,
+                    bond_label_map,
+                )
+            frame.bonds.types = bond_label_map.types
+        if self.has_angles():
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:295]: Add whitespace between to make this more readable.

File: src/lammpsio/snapshot.py [hunk 29]
```diff
@@ -487,6 +622,21 @@ def has_mass(self):
         """
         return self._mass is not None
 
+    @property
+    def type_label(self):
+        """int: LabelMap of connection types"""
+
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:628]:
```suggestion
```

File: src/lammpsio/snapshot.py [hunk 30]
```diff
@@ -487,6 +622,21 @@ def has_mass(self):
         """
         return self._mass is not None
 
+    @property
+    def type_label(self):
+        """int: LabelMap of connection types"""
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:627]:
```suggestion
        """LabelMap: Labels for particle typeids."""
```

File: src/lammpsio/topology.py [hunk 31]
```diff
@@ -150,6 +153,21 @@ def num_types(self, value):
         else:
             self._num_types = None
 
+    @property
+    def type_label(self):
+        """int: LabelMap of connection types"""
+
```
Comment (mphoward) [inline src/lammpsio/topology.py:159]:
```suggestion
        """LabelMap: Labels of connection typeids."""
```

File: src/lammpsio/topology.py [hunk 32]
```diff
@@ -241,3 +259,49 @@ class Impropers(Topology):
 
     def __init__(self, N, num_types=None):
         super().__init__(N=N, num_members=4, num_types=num_types)
+
+
+class LabelMap(collections.abc.MutableMapping):
```
Comment (mphoward) [inline src/lammpsio/topology.py:264]: This class needs documentation.

File: src/lammpsio/snapshot.py [hunk 33]
```diff
@@ -624,3 +774,29 @@ def reorder(self, order, check_order=True):
             self._charge = self._charge[order]
         if self.has_mass():
             self._mass = self._mass[order]
+
+
+def _set_type_id(lammps_typeid, gsd_typeid, label_map):
+    """Maps LAMMPS type IDs to GSD type IDs using a given label map and
+    accounting for LAMMPS being one-indexed while GSD is zero-indexed.
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:781]: Shorten this to a one-liner, potentially followed by a longer description.

File: src/lammpsio/snapshot.py [hunk 34]
```diff
@@ -487,6 +622,21 @@ def has_mass(self):
...
+    def type_label(self):
+        """int: LabelMap of connection types"""
+
+        return self._type_label
+
+    @type_label.setter
+    def type_label(self, value):
+        if value is not None:
+            if not isinstance(value, LabelMap):
+                raise TypeError("label must be a LabelMap object")
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:635]:
```suggestion
                raise TypeError("type_label must be a LabelMap")
```

File: src/lammpsio/topology.py [hunk 35]
```diff
@@ -150,6 +153,21 @@ def num_types(self, value):
...
+    def type_label(self):
+        """int: LabelMap of connection types"""
+
+        return self._type_label
+
+    @type_label.setter
+    def type_label(self, value):
+        if value is not None:
+            if not isinstance(value, LabelMap):
+                raise TypeError("label must be a LabelMap object")
```
Comment (mphoward) [inline src/lammpsio/topology.py:166]:
```suggestion
                raise TypeError("type_label must be a LabelMap")
```

**Review by mphoward — CHANGES_REQUESTED** [review]
Thanks! This is close. There is still something a little off with the particle type_label, but then this should be good to go.

File: src/lammpsio/snapshot.py (hunk 27 above)
Comment (clpetix) [reply src/lammpsio/snapshot.py:282]: I think gsd.hoomd's default behavior is for `frame.bonds.N` to be zero if undefined for all the topology. 

```python
# make a GSD frame
frame = gsd.hoomd.Frame()

frame.configuration.step = 3
frame.configuration.box = [4, 5, 6, 0.1, 0.2, 0.3]
frame.particles.N = 2
frame.particles.position = [[0.1, 0.2, 0.3], [-0.1, -0.2, -0.3]]
frame.particles.image = [[1, -1, 0], [0, 2, -2]]
frame.particles.velocity = [[1, 2, 3], [-4, -5, -6]]
frame.particles.body = [0, -1]
frame.particles.types = ["A", "B"]
frame.particles.typeid = [1, 0]
frame.particles.mass = [3, 2]
frame.particles.charge = [-1, 1]

print(frame.bonds.N)
>>> 0 
```

I think this means we should change the `from_hoomd_gsd` to be 
```python
if (
            frame.bonds.N != 0
            or frame.bonds.group is not None
            or frame.bonds.typeid is not None
            or frame.bonds.types is not None
        ):
````
and then using `self.bonds is not None` would work here

Comment (mphoward) [reply src/lammpsio/snapshot.py:282]: Yeah, that seems reasonable if the default value for that is 0 and not None! If it defaults to 0, you can also remove the bit about `is not None` when creating the topology objects.

Comment (clpetix) [reply src/lammpsio/snapshot.py:282]: > you can also remove the bit about is not None when creating the topology objects.

Comment (mphoward) [reply src/lammpsio/snapshot.py]: This part will always evaluate true now:
```
N=frame.bonds.N if frame.bonds.N is not None else 0
```
so it can be
```
N=frame.bonds.N
```

File: src/lammpsio/snapshot.py [hunk 36]
```diff
@@ -624,3 +771,37 @@ def reorder(self, order, check_order=True):
...
+    -------
+        LabelMap:
+            LabelMap mapping LAMMPS typeids to HOOMD GSD types.
+            LabelMap is created mapping typeids to str(typeids) if not provided.
+    """
+    if label_map is None:
+        sorted_typeids = numpy.sort(numpy.unique(lammps_typeid))
+        label_map = {}
+        for typeid in sorted_typeids:
+            label_map[typeid] = str(typeid)
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:798]:
```suggestion
        label_map = {typeid: str(typeid) for typeid in sorted_typeids}
```

File: src/lammpsio/snapshot.py [hunk 37]
```diff
@@ -624,3 +771,37 @@ def reorder(self, order, check_order=True):
...
+        gsd_typeid (list):
+            List of HOOMD GSD typeids to be updated (zero-indexed).
+        label_map (LabelMap):
+            LabelMap for connection type mapping LAMMPS typeids to HOOMD GSD types.
+
+    Returns:
+    -------
+        LabelMap:
+            LabelMap mapping LAMMPS typeids to HOOMD GSD types.
+            LabelMap is created mapping typeids to str(typeids) if not provided.
```
Comment (mphoward) [inline src/lammpsio/snapshot.py:792]: Remove colons after `Parameters` and `Returns` and fix indentation.

File: src/lammpsio/topology.py [hunk 38]
```diff
@@ -241,3 +259,64 @@ class Impropers(Topology):
...
+    ----------
+    map : dict
+        Map of typeids to types.
+
+    Attributes
+    ----------
+    types : tuple
+        Types in label map.
+    typeid : tuple
+        Typeids in label map.
```
Comment (mphoward) [inline src/lammpsio/topology.py:277]: These are documented as properties so you can remove this here.
```suggestion
```

File: src/lammpsio/topology.py [hunk 39]
```diff
@@ -241,3 +259,64 @@ class Impropers(Topology):
...
+    @property
+    def types(self):
+        """Types in label map.
+
+        Returns
+        -------
+        tuple
+            Label types.
+
+        """
```
Comment (mphoward) [inline src/lammpsio/topology.py:309]: Properties don't usually get a separate return since they are simple. It should go like this:
```suggestion
        """tuple: Types in label map."""
```

File: src/lammpsio/topology.py [hunk 40]
```diff
@@ -241,3 +259,64 @@ class Impropers(Topology):
...
+    @property
+    def typeid(self):
+        """Typeids in label map.
+
+        Returns
+        -------
+        tuple
+            Label typeids.
+
+        """
```
Comment (mphoward) [inline src/lammpsio/topology.py:321]:
```suggestion
        """tuple: Typeids in label map."""
```

File: tests/test_snapshot.py
Comment (mphoward) [inline tests/test_snapshot.py:1]: Please add a small test function for the `LabelMap` class itself. This will make it easier to catch underlying regressions on it in future.

**Review by mphoward — APPROVED** [review]
Looks great, thank you! Please address the small comments below and then this is ready to merge.

**Comment by clpetix** [comment]
This should be good! Thanks for your help!

**Review by mphoward — APPROVED** [review]
